### PR TITLE
[Artur Kostiuk] Uit 30 carousel

### PR DESCRIPTION
**Pull request by art1kos**
Fixes: https://issues.griddynamics.net/secure/RapidBoard.jspa?rapidView=298&view=detail&selectedIssue=UIT-30

### CODE REVIEW

File: src/app/shared/components/carousel/carousel.component.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import { ViewChild, ElementRef } from '@angular/core';
+
+@Component({
+  selector: 'app-carousel',
+  templateUrl: './carousel.component.html',
+  styleUrls: ['./carousel.component.scss']
+})
+export class CarouselComponent implements OnInit, AfterContentInit {
+
+  @ViewChild('allSlidesContainer') allSlides: ElementRef;
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.ts]: don't use an alias here. Rename to just slidesContainer


`@ViewChild()` slidesContainer: ElementRef;`

File: src/app/shared/components/carousel/carousel.component.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+
+  constructor() { }
+
+  currTranslate = 0;
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.ts]: make all properties private if they are using only inside this class

File: src/app/shared/components/carousel/carousel.component.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  childrenLength;
+
+  ngAfterContentInit() {
+    this.childrenLength = this.allSlides.nativeElement.children.length;
+  }
+  ngOnInit() {
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.ts]: according to the style guide, ngOnInit should come right before the constructor

File: src/app/shared/components/carousel/carousel.component.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    const sliderLengthCheck = (-this.childrenLength / 4 + 1) * 100;
+    if (sliderLengthCheck < this.currTranslate) {
+      this.currTranslate -= 100;
+      this.allSlides.nativeElement.style.transform = `translate(${this.currTranslate}%)`;
+    }
+  }
+  prevSlide() {
+    if (this.currTranslate) {
+      this.currTranslate += 100;
+      this.allSlides.nativeElement.style.transform = `translate(${this.currTranslate}%)`;
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.ts]: duplication logic has found. Please move to separate function

File: src/app/shared/components/carousel/carousel.component.ts (hunk 1 above)
Comment (art1kos) [reply src/app/shared/components/carousel/carousel.component.ts]: by @ViewChild() argument between '' is required.

Comment (dverbovyi) [reply src/app/shared/components/carousel/carousel.component.ts]: ok, just rename to slideContainer

File: src/app/shared/components/carousel/carousel.component.ts [hunk 5]
```diff
@@ -7,33 +7,34 @@ import { ViewChild, ElementRef } from '@angular/core';
...
-    this.allSlides.nativeElement.style.transform = `translate(${0}%)`;
+
+  private slidesMover(direction: number): void {
+    this.currTranslate += direction;
+    this.slidesContainer.nativeElement.style.transform = `translate(${this.currTranslate}%)`;
   }
 
-  nextSlide() {
-    const sliderLengthCheck = (-this.childrenLength / 4 + 1) * 100;
+  public nextSlide(): void {
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.ts]: public methods should come before private ones

File: src/app/shared/components/carousel/carousel.component.ts [hunk 6]
```diff
@@ -7,33 +7,34 @@ import { ViewChild, ElementRef } from '@angular/core';
...
-
-  ngAfterContentInit() {
-    this.childrenLength = this.allSlides.nativeElement.children.length;
+  ngAfterContentInit(): void {
+    this.childrenLength = this.slidesContainer.nativeElement.children.length;
   }
-  ngOnInit() {
-    this.allSlides.nativeElement.style.transform = `translate(${0}%)`;
+
+  private slidesMover(direction: number): void {
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.ts]: private method should come after public ones

File: src/app/home-page/home-page.module.ts [hunk 7]
```diff
@@ -3,9 +3,11 @@ import { CommonModule } from '@angular/common';
 
 import { HomePageRoutingModule } from './home-page-routing.module';
 import { HomePageComponent } from './components/home-page';
+import { CarouselComponent } from '../shared/components/carousel/carousel.component';
+
 
 @NgModule({
-  declarations: [HomePageComponent],
+  declarations: [HomePageComponent, CarouselComponent],
```
Comment (dverbovyi) [inline src/app/home-page/home-page.module.ts]: CarouselComponent is a shared component. So it should be added to the SharedModule declaration. And HomePageModule should import SharedModule only.

We talked about it yesterday

File: src/app/shared/components/carousel/carousel.component.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  styleUrls: ['./carousel.component.scss']
+})
+export class CarouselComponent implements OnInit, AfterContentInit {
+  private currTranslate = 0;
+  private childrenLength: number;
+
+  @ViewChild('slidesContainer') slidesContainer: ElementRef;
+
+  ngOnInit(): void {
+    this.slidesContainer.nativeElement.style.transform = `translate(${0}%)`;
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.ts]: the element has transtition 0 by default, so you don't need to write it here

File: src/app/shared/components/carousel/carousel.component.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  ngAfterContentInit(): void {
+    this.childrenLength = this.slidesContainer.nativeElement.children.length;
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.ts]: it's using only in 1 method, so move it there and delete ngAfterContentInit

File: src/app/shared/components/carousel/carousel.component.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  public nextSlide(): void {
+    const sliderLengthCheck: number = (-this.childrenLength / 4 + 1) * 100;
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.ts]: the variable name is not clear. The type is number but name is check?

Also, move 4 in a constant variable outside the class and give the representation name

like const ITEM_PER_PAGE = 4

wat8

File: src/app/shared/components/carousel/carousel.component.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    if (sliderLengthCheck < this.currTranslate) {
+      this.slidesMover(-100);
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.ts]: the function has to have verb in the name

**Comment by dverbovyi** [comment]
@art1kos once you done with requested changes, please fix the merge conflicts

File: src/app/shared/components/carousel/carousel.component.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    }
+  }
+
+  public prevSlide(): void {
+    if (this.currTranslate) {
+      this.moveSlides(100 / this.childrenLength * this.itemsPerPage);
+    }
+  }
+
+  private moveSlides(direction: number): void {
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.ts]: rename direction to translateStep

File: src/app/shared/components/carousel/carousel.component.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+      this.itemsPerPage = 3;
+    } else {
+      this.itemWidth = '250%';
+      this.itemsPerPage = 4;
+    }
+  }
+
+  ngOnInit(): void {
+    this.sizeHandler();
+    const event = fromEvent(window, 'resize').subscribe(evt => {
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.ts]: 1 unsubscribe in ngOnDestory
2 add debounceTime ask @taras1313 for clarification

**Comment by dverbovyi** [comment]
@art1kos I'm merging this, but please create separate MR with unit-test ASAP

**Pull request merged**